lib/fleek-ipld/src/walker/stream.rs

```diff
@@ -0,0 +1,98 @@
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+use futures::future::BoxFuture;
+use futures::TryFuture;
+use tokio_stream::Stream;
+
+use super::downloader::Downloader;
+use crate::decoder::data_codec::Decoder;
+use crate::decoder::fs::{DirItem, DocId, IpldItem, Link};
+use crate::decoder::reader::IpldReader;
+use crate::errors::IpldError;
+
+pub struct IpldStream<D, C> {
+    downloader: D,
+    reader: IpldReader<C>,
+    stack: Vec<(Option<DirItem>, Vec<Link>, usize)>,
+    pending_future: Option<BoxFuture<'static, Result<Option<IpldItem>, IpldError>>>,
+}
+
+impl<D, C> IpldStream<D, C>
+where
+    D: Downloader + Clone + Send + Sync + 'static,
+    C: Decoder + Clone + Send + Sync + 'static,
+{
+    pub fn new(downloader: D, reader: IpldReader<C>, id: DocId) -> Self {
+        Self {
+            downloader,
+            reader,
+            stack: vec![(None, vec![id.into()], 0)],
+            pending_future: None,
+        }
+    }
+}
+
+/// Implementing Stream for IpldStream in order to traverse all the links associated to a given Cid
+impl<D, C> Stream for IpldStream<D, C>
+where
+    D: Downloader + Clone + Unpin + Send + Sync + 'static,
+    C: Decoder + Clone + Unpin + Send + Sync + 'static,
+{
+    type Item = Result<IpldItem, IpldError>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.get_mut();
+
+        if let Some(fut) = this.pending_future.as_mut() {
+            match Pin::new(fut).try_poll(cx) {
+                Poll::Ready(result) => {
+                    this.pending_future = None;
+                    match result {
+                        Ok(Some(ref i @ IpldItem::Dir(ref item))) => {
+                            let links = item.links().clone();
+                            this.stack.push((Some(item.clone()), links, 0));
+                            return Poll::Ready(Some(Ok(i.clone())));
+                        },
+                        Ok(Some(item @ IpldItem::File(_))) => {
+                            return Poll::Ready(Some(Ok(item)));
+                        },
+                        Ok(Some(chunk @ IpldItem::ChunkedFile(_))) => {
+                            return Poll::Ready(Some(Ok(chunk)));
+                        },
```

Comment:
maybe:
```suggestion
                        Ok(Some(item)) => {
                            return Poll::Ready(Some(Ok(item)));
                        },
```

Reply:
This file was deleted ... 😄  I changed for other impl